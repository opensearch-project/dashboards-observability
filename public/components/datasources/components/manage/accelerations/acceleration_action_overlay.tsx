/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useState } from 'react';
import { EuiOverlayMask, EuiConfirmModal, EuiCompressedFormRow, EuiCompressedFieldText } from '@elastic/eui';
import { CachedAcceleration } from '../../../../../../common/types/data_connections';
import {
  ACC_DELETE_MSG,
  ACC_VACUUM_MSG,
  ACC_SYNC_MSG,
  AccelerationActionType,
  getAccelerationName,
  getAccelerationFullPath,
} from './utils/acceleration_utils';

export interface AccelerationActionOverlayProps {
  isVisible: boolean;
  actionType: AccelerationActionType;
  acceleration: CachedAcceleration | null;
  dataSourceName: string;
  onCancel: () => void;
  onConfirm: () => void;
}

export const AccelerationActionOverlay: React.FC<AccelerationActionOverlayProps> = ({
  isVisible,
  actionType,
  acceleration,
  dataSourceName,
  onCancel,
  onConfirm,
}) => {
  const [confirmationInput, setConfirmationInput] = useState('');

  if (!isVisible || !acceleration) {
    return null;
  }

  const displayIndexName = getAccelerationName(acceleration);
  const displayFullPath = getAccelerationFullPath(acceleration, dataSourceName);

  let title = '';
  let description = '';
  let confirmButtonText = 'Confirm';
  let confirmEnabled = true;

  switch (actionType) {
    case 'vacuum':
      title = `Vacuum acceleration ${displayIndexName} on ${displayFullPath} ?`;
      description = ACC_VACUUM_MSG;
      confirmButtonText = 'Vacuum';
      confirmEnabled = confirmationInput === displayIndexName;
      break;
    case 'delete':
      title = `Delete acceleration ${displayIndexName} on ${displayFullPath} ?`;
      description = ACC_DELETE_MSG(displayIndexName);
      confirmButtonText = 'Delete';
      break;
    case 'sync':
      title = 'Manual sync data?';
      description = ACC_SYNC_MSG;
      confirmButtonText = 'Sync';
      break;
  }

  return (
    <EuiOverlayMask>
      <EuiConfirmModal
        title={title}
        onCancel={onCancel}
        onConfirm={() => onConfirm()}
        cancelButtonText="Cancel"
        confirmButtonText={confirmButtonText}
        buttonColor="danger"
        defaultFocusedButton="confirm"
        confirmButtonDisabled={!confirmEnabled}
      >
        <p>{description}</p>
        {actionType === 'vacuum' && (
          <EuiCompressedFormRow label={`To confirm, type ${displayIndexName}`}>
            <EuiCompressedFieldText
              name="confirmationInput"
              value={confirmationInput}
              onChange={(e) => setConfirmationInput(e.target.value)}
            />
          </EuiCompressedFormRow>
        )}
      </EuiConfirmModal>
    </EuiOverlayMask>
  );
};
