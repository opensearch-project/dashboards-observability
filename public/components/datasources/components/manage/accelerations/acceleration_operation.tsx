/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { useEffect, useState } from 'react';
import { CachedAcceleration } from '../../../../../../common/types/data_connections';
import { useToast } from '../../../../common/toast';
import { useDirectQuery } from '../../../../../framework/datasources/direct_query_hook';
import { DirectQueryLoadingStatus } from '../../../../../../common/types/explorer';
import {
  AccelerationActionType,
  generateAccelerationOperationQuery,
  getAccelerationName,
} from './utils/acceleration_utils';

export const useAccelerationOperation = (dataSource: string) => {
  const { startLoading, stopLoading, loadStatus } = useDirectQuery();
  const { setToast } = useToast();
  const [isOperating, setIsOperating] = useState(false);
  const [operationSuccess, setOperationSuccess] = useState(false);
  const [accelerationToOperate, setAccelerationToOperate] = useState<CachedAcceleration | null>(
    null
  );
  const [operationType, setOperationType] = useState<AccelerationActionType | null>(null);
  const [currentStatus, setCurrentStatus] = useState<DirectQueryLoadingStatus | null>(null);

  useEffect(() => {
    if (!accelerationToOperate || !operationType || loadStatus === currentStatus) return;

    const displayAccelerationName = getAccelerationName(accelerationToOperate);

    let operationInProgressMessage = '';
    let operationSuccessMessage = '';
    let operationFailureMessage = '';

    switch (operationType) {
      case 'delete':
        operationInProgressMessage = `Deleting acceleration: ${displayAccelerationName}`;
        operationSuccessMessage = `Successfully deleted acceleration: ${displayAccelerationName}`;
        operationFailureMessage = `Failed to delete acceleration: ${displayAccelerationName}`;
        break;
      case 'vacuum':
        operationInProgressMessage = `Vacuuming acceleration: ${displayAccelerationName}`;
        operationSuccessMessage = `Successfully vacuumed acceleration: ${displayAccelerationName}`;
        operationFailureMessage = `Failed to vacuum acceleration: ${displayAccelerationName}`;
        break;
      case 'sync':
        operationInProgressMessage = `Syncing acceleration: ${displayAccelerationName}`;
        break;
    }

    if (loadStatus === DirectQueryLoadingStatus.SCHEDULED && operationType !== 'sync') {
      setIsOperating(true);
      setToast(operationInProgressMessage, 'success');
    } else if (loadStatus === DirectQueryLoadingStatus.SUCCESS && operationType !== 'sync') {
      setIsOperating(false);
      setAccelerationToOperate(null);
      setOperationSuccess(true);
      setToast(operationSuccessMessage, 'success');
    } else if (loadStatus === DirectQueryLoadingStatus.FAILED && operationType !== 'sync') {
      setIsOperating(false);
      setOperationSuccess(false);
      setToast(operationFailureMessage, 'danger');
    } else if (operationType === 'sync' && loadStatus === DirectQueryLoadingStatus.SCHEDULED) {
      setToast(operationInProgressMessage, 'success');
      stopLoading();
    }

    setCurrentStatus(loadStatus);
  }, [loadStatus, setToast, accelerationToOperate, operationType, currentStatus]);

  useEffect(() => {
    return () => {
      stopLoading();
    };
  }, []);

  const performOperation = (
    acceleration: CachedAcceleration,
    operation: AccelerationActionType
  ) => {
    setOperationSuccess(false);
    setOperationType(operation);
    const operationQuery = generateAccelerationOperationQuery(acceleration, dataSource, operation);

    const requestPayload = {
      lang: 'sql',
      query: operationQuery,
      datasource: dataSource,
    };

    setIsOperating(true);
    setAccelerationToOperate(acceleration);
    startLoading(requestPayload);
  };

  return { performOperation, isOperating, operationSuccess };
};
