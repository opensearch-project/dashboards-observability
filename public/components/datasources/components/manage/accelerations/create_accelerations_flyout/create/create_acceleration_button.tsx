/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiButton } from '@elastic/eui';
import React, { useEffect, useState } from 'react';
import { CreateAccelerationForm } from '../../../../../../../../common/types/data_connections';
import {
  DirectQueryLoadingStatus,
  DirectQueryRequest,
} from '../../../../../../../../common/types/explorer';
import { useDirectQuery } from '../../../../../../../framework/datasources/direct_query_hook';
import { useToast } from '../../../../../../common/toast';
import { accelerationQueryBuilder } from '../visual_editors/query_builder';
import { formValidator, hasError } from './utils';

interface CreateAccelerationButtonProps {
  accelerationFormData: CreateAccelerationForm;
  setAccelerationFormData: React.Dispatch<React.SetStateAction<CreateAccelerationForm>>;
  resetFlyout: () => void;
}

export const CreateAccelerationButton = ({
  accelerationFormData,
  setAccelerationFormData,
  resetFlyout,
}: CreateAccelerationButtonProps) => {
  const { setToast } = useToast();
  const { loadStatus: directqueryLoadStatus, startLoading: startDirectQuery } = useDirectQuery();
  const [isLoading, setIsLoading] = useState(false);

  const createAcceleration = () => {
    const errors = formValidator(accelerationFormData);
    if (hasError(errors)) {
      setAccelerationFormData({ ...accelerationFormData, formErrors: errors });
      return;
    }

    const requestPayload: DirectQueryRequest = {
      lang: 'sql',
      query: accelerationQueryBuilder(accelerationFormData),
      datasource: accelerationFormData.dataSource,
    };

    startDirectQuery(requestPayload);
    setIsLoading(true);
  };

  useEffect(() => {
    const status = directqueryLoadStatus.toLowerCase();
    if (status === DirectQueryLoadingStatus.SUCCESS) {
      setIsLoading(false);
      setToast('Create acceleration query submitted successfully!', 'success');
      resetFlyout();
    } else if (
      status === DirectQueryLoadingStatus.FAILED ||
      status === DirectQueryLoadingStatus.CANCELED
    ) {
      setIsLoading(false);
      setToast('Create acceleration query failed', 'success');
    }
  }, [directqueryLoadStatus]);

  return (
    <EuiButton onClick={createAcceleration} fill isLoading={isLoading}>
      Create acceleration
    </EuiButton>
  );
};
