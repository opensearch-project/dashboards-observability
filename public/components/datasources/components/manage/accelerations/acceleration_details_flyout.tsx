/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiButtonEmpty,
  EuiFlexGroup,
  EuiFlexItem,
  EuiFlyoutBody,
  EuiFlyoutHeader,
  EuiIcon,
  EuiSpacer,
  EuiTab,
  EuiTabs,
  EuiText,
} from '@elastic/eui';
import React, { useEffect, useState } from 'react';
import { OpenSearchDashboardsResponse } from '../../../../../../../../src/core/server/http/router';
import { CachedAcceleration } from '../../../../../../common/types/data_connections';
import { coreRefs } from '../../../../../framework/core_refs';
import { AccelerationActionOverlay } from './acceleration_action_overlay';
import { useAccelerationOperation } from './acceleration_operation';
import { AccelerationDetailsTab } from './flyout_modules/acceleration_details_tab';
import { AccelerationSchemaTab } from './flyout_modules/accelerations_schema_tab';
import {
  AccelerationActionType,
  getAccelerationName,
  onDiscoverIconClick,
} from './utils/acceleration_utils';

export interface AccelerationDetailsFlyoutProps {
  acceleration: CachedAcceleration;
  dataSourceName: string;
  resetFlyout: () => void;
  handleRefresh?: () => void;
  dataSourceMDSId?: string;
}

const getMappings = (index: string): Promise<OpenSearchDashboardsResponse> | undefined => {
  return coreRefs.dslService?.fetchFields(index);
};

const getSettings = (index: string): Promise<OpenSearchDashboardsResponse> | undefined => {
  return coreRefs.dslService?.fetchSettings(index);
};

const getIndexInfo = (index: string): Promise<OpenSearchDashboardsResponse> | undefined => {
  return coreRefs.dslService?.fetchIndices(index);
};

const handleDetailsFetchingPromise = (
  promise: Promise<OpenSearchDashboardsResponse> | undefined,
  action: string
) => {
  return promise!
    .then((data) => ({ status: 'fulfilled', action, data }))
    .catch((error) => ({ status: 'rejected', action, error }));
};

export const AccelerationDetailsFlyout = (props: AccelerationDetailsFlyoutProps) => {
  const { dataSourceName, acceleration, resetFlyout, handleRefresh } = props;
  const { flintIndexName } = acceleration;
  const [selectedTab, setSelectedTab] = useState('details');
  const tabsMap: { [key: string]: any } = {
    details: AccelerationDetailsTab,
    schema: AccelerationSchemaTab,
  };
  const [operationType, setOperationType] = useState<AccelerationActionType | null>(null);
  const [showConfirmationOverlay, setShowConfirmationOverlay] = useState(false);

  const { performOperation, operationSuccess } = useAccelerationOperation(props.dataSourceName);

  const displayedIndex = getAccelerationName(acceleration);

  const onConfirmOperation = () => {
    if (operationType && props.acceleration) {
      performOperation(props.acceleration, operationType);
      setShowConfirmationOverlay(false);
    }
  };

  const onSyncIconClickHandler = () => {
    setOperationType('sync');
    setShowConfirmationOverlay(true);
  };

  const onDeleteIconClickHandler = () => {
    setOperationType('delete');
    setShowConfirmationOverlay(true);
  };

  const onVacuumIconClickHandler = () => {
    setOperationType('vacuum');
    setShowConfirmationOverlay(true);
  };

  const [settings, setSettings] = useState<object>();
  const [mappings, setMappings] = useState();
  const [indexInfo, setIndexInfo] = useState();

  const updateMapping = (result) => {
    setMappings(result);
  };

  const updateSetting = (result, slectedIndex: string) => {
    setSettings(result.data[slectedIndex]);
  };

  const updateIndexInfo = (result) => {
    setIndexInfo(result);
  };

  const getAccDetail = (selectedIndex: string) => {
    Promise.all([
      handleDetailsFetchingPromise(getMappings(selectedIndex), 'getMappings'),
      handleDetailsFetchingPromise(getSettings(selectedIndex), 'getSettings'),
      handleDetailsFetchingPromise(getIndexInfo(selectedIndex), 'getIndexInfo'),
    ])
      .then((results) => {
        updateMapping(results[0]);
        updateSetting(results[1], selectedIndex);
        updateIndexInfo(results[2]);
      })
      .catch((errors: Error[]) => {
        errors.forEach((error, errorIndex) => {
          console.error(`Error in async call ${errorIndex + 1}:`, error);
        });
      });
  };

  useEffect(() => {
    if (operationSuccess) {
      resetFlyout();
      handleRefresh?.();
      setOperationType(null);
      setShowConfirmationOverlay(false);
    }
  }, [operationSuccess, resetFlyout, handleRefresh]);

  useEffect(() => {
    if (flintIndexName !== undefined && flintIndexName.trim().length > 0) {
      getAccDetail(flintIndexName);
    }
  }, [flintIndexName]);

  const DiscoverIcon = () => {
    return (
      <EuiButtonEmpty
        onClick={() => {
          onDiscoverIconClick(acceleration, dataSourceName);
          resetFlyout();
        }}
      >
        <EuiIcon type={'discoverApp'} size="m" />
      </EuiButtonEmpty>
    );
  };

  const SyncIcon = ({ autoRefresh, status }: { autoRefresh: boolean; status: string }) => {
    if (autoRefresh || status !== 'active') {
      return null;
    }
    return (
      <EuiButtonEmpty onClick={onSyncIconClickHandler}>
        <EuiIcon type="inputOutput" size="m" />
      </EuiButtonEmpty>
    );
  };

  const DeleteIcon = () => {
    return (
      <EuiButtonEmpty onClick={onDeleteIconClickHandler}>
        <EuiIcon type="trash" size="m" />
      </EuiButtonEmpty>
    );
  };

  const VacuumIcon = () => {
    return (
      <EuiButtonEmpty onClick={onVacuumIconClickHandler}>
        <EuiIcon type="broom" size="m" />
      </EuiButtonEmpty>
    );
  };

  const accelerationDetailsTabs = [
    {
      id: 'details',
      name: 'Details',
      disabled: false,
    },
    {
      id: 'schema',
      name: 'Schema',
      disabled: false,
    },
  ];

  const renderTabs = () => {
    return accelerationDetailsTabs.map((tab, tabIndex) => {
      return (
        <EuiTab
          onClick={() => setSelectedTab(tab.id)}
          isSelected={tab.id === selectedTab}
          disabled={tab.disabled}
          key={tabIndex}
        >
          {tab.name}
        </EuiTab>
      );
    });
  };

  const renderTabContent = (tab: string) => {
    let propsForTab;

    switch (tab) {
      case 'details':
        propsForTab = { acceleration, settings, mappings, indexInfo, dataSourceName, resetFlyout };
        break;
      case 'schema':
        propsForTab = { mappings, indexInfo };
        break;
      default:
        return null;
    }

    const TabToDisplay = tabsMap[tab];

    return <TabToDisplay {...propsForTab} />;
  };

  return (
    <>
      <EuiFlyoutHeader hasBorder>
        <EuiFlexGroup direction="row" alignItems="center" gutterSize="m">
          <EuiFlexItem>
            <EuiText>
              <h2 className="panel-title">{displayedIndex}</h2>
            </EuiText>
          </EuiFlexItem>
          <EuiFlexItem grow={false}>
            <SyncIcon autoRefresh={acceleration.autoRefresh} status={acceleration.status} />
          </EuiFlexItem>
          <EuiFlexItem grow={false}>
            <DiscoverIcon />
          </EuiFlexItem>
          {acceleration.status !== 'deleted' ? (
            <EuiFlexItem grow={false}>
              <DeleteIcon />
            </EuiFlexItem>
          ) : (
            <EuiFlexItem grow={false}>
              <VacuumIcon />
            </EuiFlexItem>
          )}
        </EuiFlexGroup>
        <EuiSpacer size="m" />
        <EuiTabs style={{ marginBottom: '-25px' }}>{renderTabs()}</EuiTabs>
      </EuiFlyoutHeader>
      <EuiFlyoutBody>{renderTabContent(selectedTab)}</EuiFlyoutBody>
      {showConfirmationOverlay && operationType && (
        <AccelerationActionOverlay
          isVisible={showConfirmationOverlay}
          actionType={operationType as AccelerationActionType}
          acceleration={acceleration}
          dataSourceName={dataSourceName}
          onCancel={() => setShowConfirmationOverlay(false)}
          onConfirm={onConfirmOperation}
        />
      )}
    </>
  );
};
