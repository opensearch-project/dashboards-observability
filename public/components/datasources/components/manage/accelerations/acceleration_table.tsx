/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiButton,
  EuiEmptyPrompt,
  EuiFlexGroup,
  EuiFlexItem,
  EuiHorizontalRule,
  EuiInMemoryTable,
  EuiLink,
  EuiLoadingSpinner,
  EuiPanel,
  EuiSpacer,
  EuiTableFieldDataColumnType,
  EuiText,
} from '@elastic/eui';
import React, { useCallback, useEffect, useState } from 'react';
import {
  CachedAcceleration,
  CachedDataSourceStatus,
} from '../../../../../../common/types/data_connections';
import { DirectQueryLoadingStatus } from '../../../../../../common/types/explorer';
import { CatalogCacheManager } from '../../../../../framework/catalog_cache/cache_manager';
import {
  getRenderAccelerationDetailsFlyout,
  getRenderCreateAccelerationFlyout,
} from '../../../../../plugin';
import { isCatalogCacheFetching } from '../associated_objects/utils/associated_objects_tab_utils';
import { AccelerationActionOverlay } from './acceleration_action_overlay';
import { useAccelerationOperation } from './acceleration_operation';
import {
  ACC_LOADING_MSG,
  ACC_PANEL_DESC,
  ACC_PANEL_TITLE,
  AccelerationActionType,
  AccelerationStatus,
  CreateAccelerationFlyoutButton,
  getAccelerationName,
  onDiscoverIconClick,
} from './utils/acceleration_utils';

interface AccelerationTableProps {
  dataSourceName: string;
  cacheLoadingHooks: any;
}

interface ModalState {
  actionType: AccelerationActionType | null;
  selectedItem: CachedAcceleration | null;
}

export const AccelerationTable = ({
  dataSourceName,
  cacheLoadingHooks,
}: AccelerationTableProps) => {
  const [accelerations, setAccelerations] = useState<CachedAcceleration[]>([]);
  const [updatedTime, setUpdatedTime] = useState<string>();
  const { performOperation, operationSuccess } = useAccelerationOperation(dataSourceName);
  const {
    databasesLoadStatus,
    tablesLoadStatus,
    accelerationsLoadStatus,
    startLoadingAccelerations,
  } = cacheLoadingHooks;
  const [isRefreshing, setIsRefreshing] = useState(false);
  const [modalState, setModalState] = useState<ModalState>({
    actionType: null,
    selectedItem: null,
  });

  useEffect(() => {
    if (operationSuccess) {
      handleRefresh();
    }
  }, [operationSuccess]);

  const handleActionClick = (
    actionType: ModalState['actionType'],
    acceleration: CachedAcceleration
  ) => {
    setModalState({
      actionType,
      selectedItem: acceleration,
    });
  };

  const handleModalClose = () => {
    setModalState({
      actionType: null,
      selectedItem: null,
    });
  };

  const handleConfirm = () => {
    if (!modalState.selectedItem || !modalState.actionType) return;

    performOperation(modalState.selectedItem, modalState.actionType);
    handleModalClose();
  };

  useEffect(() => {
    const cachedDataSource = CatalogCacheManager.getOrCreateAccelerationsByDataSource(
      dataSourceName
    );
    if (
      cachedDataSource.status === CachedDataSourceStatus.Empty &&
      !isCatalogCacheFetching(accelerationsLoadStatus)
    ) {
      setIsRefreshing(true);
      startLoadingAccelerations({ dataSourceName });
    } else {
      setAccelerations(cachedDataSource.accelerations);
      setUpdatedTime(cachedDataSource.lastUpdated);
    }
  }, []);

  useEffect(() => {
    if (accelerationsLoadStatus === DirectQueryLoadingStatus.SUCCESS) {
      const cachedDataSource = CatalogCacheManager.getOrCreateAccelerationsByDataSource(
        dataSourceName
      );
      setAccelerations(cachedDataSource.accelerations);
      setUpdatedTime(cachedDataSource.lastUpdated);
      setIsRefreshing(false);
    }
    if (accelerationsLoadStatus === DirectQueryLoadingStatus.FAILED) {
      setIsRefreshing(false);
    }
  }, [accelerationsLoadStatus]);

  const handleRefresh = useCallback(() => {
    if (!isCatalogCacheFetching(accelerationsLoadStatus)) {
      setIsRefreshing(true);
      startLoadingAccelerations(dataSourceName);
    }
  }, [accelerationsLoadStatus]);

  const RefreshButton = () => {
    return (
      <EuiButton
        onClick={handleRefresh}
        isLoading={
          isRefreshing ||
          isCatalogCacheFetching(databasesLoadStatus, tablesLoadStatus, accelerationsLoadStatus)
        }
      >
        Refresh
      </EuiButton>
    );
  };

  const displayUpdatedTime = updatedTime ? new Date(updatedTime).toLocaleString() : '';

  const AccelerationTableHeader = () => {
    return (
      <>
        <EuiFlexGroup direction="row" alignItems="center">
          <EuiFlexItem>
            <EuiText>
              <h2 className="panel-title">{ACC_PANEL_TITLE}</h2>
              {ACC_PANEL_DESC}
            </EuiText>
          </EuiFlexItem>
          <EuiFlexItem grow={false}>
            <EuiFlexGroup direction="rowReverse" alignItems="flexEnd">
              <EuiFlexItem grow={false}>
                <CreateAccelerationFlyoutButton
                  dataSourceName={dataSourceName}
                  renderCreateAccelerationFlyout={renderCreateAccelerationFlyout}
                  handleRefresh={handleRefresh}
                />
              </EuiFlexItem>
              <EuiFlexItem grow={false}>
                <RefreshButton data-test-subj="refreshButton" />
              </EuiFlexItem>
              {updatedTime && (
                <EuiFlexItem>
                  <EuiText textAlign="right" size="xs" color="subdued">
                    {'Last updated at:'}
                  </EuiText>
                  <EuiText textAlign="right" color="subdued" size="xs">
                    {displayUpdatedTime}
                  </EuiText>
                </EuiFlexItem>
              )}
            </EuiFlexGroup>
          </EuiFlexItem>
        </EuiFlexGroup>
      </>
    );
  };

  const AccelerationLoading = () => {
    const BodyText = () => (
      <>
        <p>{ACC_LOADING_MSG}</p>
      </>
    );

    return <EuiEmptyPrompt icon={<EuiLoadingSpinner size="xl" />} body={<BodyText />} />;
  };

  const tableActions = [
    {
      name: 'Query Data',
      description: 'Query in Observability Logs',
      icon: 'discoverApp',
      type: 'icon',
      onClick: (acc: CachedAcceleration) => {
        onDiscoverIconClick(acc, dataSourceName);
      },
    },
    {
      name: 'Sync',
      description: 'Manual Sync Data',
      icon: 'inputOutput',
      onClick: (item: CachedAcceleration) => handleActionClick('sync', item),
      enabled: (item: CachedAcceleration) => !item.autoRefresh && item.status === 'active',
    },
    {
      name: 'Delete',
      description: 'Delete acceleration',
      icon: 'trash',
      onClick: (item: CachedAcceleration) => handleActionClick('delete', item),
      enabled: (item: CachedAcceleration) => item.status !== 'deleted',
    },
    {
      name: 'Vacuum',
      description: 'Vacuum acceleration',
      icon: 'broom',
      onClick: (item: CachedAcceleration) => handleActionClick('vacuum', item),
      enabled: (item: CachedAcceleration) => item.status === 'deleted',
    },
  ];

  const accelerationTableColumns = [
    {
      field: 'indexName',
      name: 'Name',
      sortable: true,
      render: (indexName: string, acceleration: CachedAcceleration) => {
        const displayName = getAccelerationName(acceleration);
        return (
          <EuiLink
            onClick={() => {
              renderAccelerationDetailsFlyout(acceleration, dataSourceName, handleRefresh);
            }}
          >
            {displayName}
          </EuiLink>
        );
      },
    },
    {
      field: 'status',
      name: 'Status',
      sortable: true,
      render: (status: string) => <AccelerationStatus status={status} />,
    },
    {
      field: 'type',
      name: 'Type',
      sortable: true,
      render: (type: string) => {
        let label;
        switch (type) {
          case 'skipping':
            label = 'Skipping Index';
            break;
          case 'materialized':
            label = 'Materialized View';
            break;
          case 'covering':
            label = 'Covering Index';
            break;
          default:
            label = 'INVALID TYPE';
        }
        return <EuiText size="s">{label}</EuiText>;
      },
    },
    {
      field: 'database',
      name: 'Database',
      sortable: true,
      render: (database: string) => <EuiText size="s">{database}</EuiText>,
    },
    {
      field: 'table',
      name: 'Table',
      sortable: true,
      render: (table: string) => <EuiText size="s">{table || '-'}</EuiText>,
    },
    {
      field: 'refreshType',
      name: 'Refresh Type',
      sortable: true,
      render: (autoRefresh: boolean, acceleration: CachedAcceleration) => {
        return <EuiText size="s">{acceleration.autoRefresh ? 'Auto refresh' : 'Manual'}</EuiText>;
      },
    },
    {
      field: 'flintIndexName',
      name: 'Destination Index',
      sortable: true,
      render: (flintIndexName: string, acceleration: CachedAcceleration) => {
        if (acceleration.type === 'skipping') {
          return '-';
        }
        return flintIndexName || '-';
      },
    },
    {
      name: 'Actions',
      actions: tableActions,
    },
  ] as Array<EuiTableFieldDataColumnType<any>>;

  const pagination = {
    initialPageSize: 10,
    pageSizeOptions: [10, 20, 50, 100],
  };

  const sorting = {
    sort: {
      field: 'name',
      direction: 'asc',
    },
  };

  const renderAccelerationDetailsFlyout = getRenderAccelerationDetailsFlyout();
  const renderCreateAccelerationFlyout = getRenderCreateAccelerationFlyout();

  return (
    <>
      <EuiSpacer />
      <EuiPanel>
        <AccelerationTableHeader />
        <EuiHorizontalRule />
        {isRefreshing ? (
          <AccelerationLoading />
        ) : (
          <EuiInMemoryTable
            items={accelerations}
            columns={accelerationTableColumns}
            pagination={pagination}
            sorting={sorting}
          />
        )}
      </EuiPanel>
      {(modalState.actionType === 'delete' ||
        modalState.actionType === 'vacuum' ||
        modalState.actionType === 'sync') && (
        <AccelerationActionOverlay
          isVisible={!!modalState.actionType}
          actionType={modalState.actionType}
          acceleration={modalState.selectedItem}
          dataSourceName={dataSourceName}
          onCancel={handleModalClose}
          onConfirm={handleConfirm}
        />
      )}
    </>
  );
};
