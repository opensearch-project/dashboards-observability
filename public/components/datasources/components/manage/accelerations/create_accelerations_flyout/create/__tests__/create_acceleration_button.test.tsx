/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { waitFor } from '@testing-library/dom';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import React from 'react';
import { materializedViewBuilderMock2 } from '../../../../../../../../../test/accelerations';
import { CreateAccelerationButton } from '../create_acceleration_button';

describe('Create acceleration button component', () => {
  configure({ adapter: new Adapter() });

  it('renders create acceleration button component with mv state', async () => {
    const accelerationFormData = materializedViewBuilderMock2;
    const setAccelerationFormData = jest.fn();
    const wrapper = mount(
      <CreateAccelerationButton
        accelerationFormData={accelerationFormData}
        setAccelerationFormData={setAccelerationFormData}
        resetFlyout={jest.fn()}
      />
    );
    wrapper.update();
    await waitFor(() => {
      expect(
        toJson(wrapper, {
          noKey: false,
          mode: 'deep',
        })
      ).toMatchSnapshot();
    });
  });
});
