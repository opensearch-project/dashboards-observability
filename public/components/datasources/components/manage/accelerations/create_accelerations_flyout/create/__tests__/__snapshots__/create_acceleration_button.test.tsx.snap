// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create acceleration button component renders create acceleration button component with mv state 1`] = `
<button
  className="euiButton euiButton--primary euiButton--fill"
  disabled={false}
  onClick={[Function]}
  style={
    Object {
      "minWidth": undefined,
    }
  }
  type="button"
>
  <span
    className="euiButtonContent euiButton__content"
  >
    <span
      className="euiButton__text"
    >
      Create acceleration
    </span>
  </span>
</button>
`;
